File: app/src/main/java/com/shen/xi/android/tut/WeiboClientManager.java
package com.shen.xi.android.tut;


import android.os.AsyncTask;

import com.google.common.collect.Collections2;
import com.google.common.collect.Lists;
import com.google.inject.Inject;
import com.squareup.otto.Bus;
import com.squareup.otto.Subscribe;
import com.shen.xi.android.tut.event.RefreshStatusCompleteEvent;
import com.shen.xi.android.tut.event.RefreshWeiboEvent;
import com.shen.xi.android.tut.util.Util;
import com.shen.xi.android.tut.weibo.WeiboClient;
import com.shen.xi.android.tut.weibo.WeiboStatus;
import com.shen.xi.android.tut.weibo.WeiboTimeline;
import com.shen.xi.android.tut.weibo.WeiboTimelineException;

import java.util.Collection;
import java.util.List;

public class WeiboClientManager {

    private static final String TAG = WeiboClientManager.class.getName();
    private final WeiboClient mClient;
    private final Bus mBus;

    @Inject
    public WeiboClientManager(Bus bus, WeiboClient client) {
        mBus = bus;
        mClient = client;

        mBus.register(this);
    }

    @Subscribe
    public void refreshStatus(RefreshWeiboEvent event) {
        new AsyncTask<String, Integer, List<WeiboStatus>>() {

            private final RefreshStatusCompleteEvent completeEvent = new RefreshStatusCompleteEvent();

            @Override
            protected List<WeiboStatus> doInBackground(String... strings) {
                List<WeiboStatus> statusList = null;

                try {
                    String sinceId = strings[0];
                    WeiboTimeline timeline = mClient.getHomeTimeline(sinceId);

                    if (timeline != null && timeline.statuses.size() > 0) {
                        Collection<WeiboStatus> statusCollection =
                                Collections2.filter(timeline.statuses, Util.ImageUrlPredictor);
                        statusList = Lists.newArrayList(statusCollection);
                    }
                } catch (WeiboTimelineException e) {
                    e.printStackTrace();
                }

                return statusList;
            }

            @Override
            protected void onPostExecute(List<WeiboStatus> weiboStatuses) {
                mBus.post(completeEvent.setStatus(weiboStatuses));
            }

        }.execute(event.sinceId);
    }

    @Override
    protected void finalize() throws Throwable {
        mBus.unregister(this);

        super.finalize();
    }

}

File: app/src/main/scala/com/shen/xi/android/tut/WeiboClientManager.scala
package com.shen.xi.android.tut


import com.google.inject.Inject
import com.squareup.otto.{Bus, Subscribe}

import com.shen.xi.android.tut.event.{RefreshStatusCompleteEvent, RefreshWeiboEvent}
import com.shen.xi.android.tut.weibo.WeiboClient

import java.lang.Runnable

import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.util.{Success, Failure}
import scala.collection.JavaConversions._


class WeiboClientManager @Inject()(bus: Bus, client: WeiboClient) {

  private val TAG = classOf[WeiboClientManager].getName
  private val mClient = client
  private val mBus = bus

  mBus.register(this)

  @Subscribe
  def refreshStatus(event: RefreshWeiboEvent): Unit = {
    val completeEvent = new RefreshStatusCompleteEvent()

    Future {

      val timeline = mClient.getHomeTimeline(event.sinceId)
      if (timeline != null && timeline.statuses.size > 0) {
        bufferAsJavaList(timeline.statuses filter(_.getImageUrl != null) distinct)
      } else null

    } onComplete {

      case Success(statuses) => {
        event.activity.runOnUiThread(new Runnable() {
          def run() {
            mBus.post(completeEvent.setStatus(statuses))
          }
        })
      }

      case Failure(e) => {
        e.printStackTrace()

        event.activity.runOnUiThread(new Runnable() {
          def run() {
            mBus.post(completeEvent.setStatus(null))
          }
        })
      }

    }
  }

  @throws[Throwable]
  override protected def finalize() = {
    mBus.unregister(this)

    super.finalize()
  }

}
